keyserver/listener.go

```diff
@@ -0,0 +1,108 @@
+package keyserver
+
+import (
+	"crypto/tls"
+	"io"
+	"log"
+	"net"
+	"time"
+
+	. "github.com/coniks-sys/coniks-go/protocol"
+)
+
+func (server *ConiksServer) listenForRequests(ln *net.TCPListener, handler func(msg []byte) ([]byte, error)) {
+	defer ln.Close()
+	for {
+		select {
+		case <-server.stop:
+			server.waitCloseConn.Wait()
+			return
+		}
+		ln.SetDeadline(time.Now().Add(5 * time.Second))
+		conn, err := ln.Accept()
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok && opErr.Timeout() {
+				continue
+			}
+			log.Printf("accept client: %s", err)
+			continue
+		}
+		conn = tls.Server(conn, server.tlsConfig)
+		server.waitCloseConn.Add(1)
+		go func() {
+			server.acceptClient(conn, handler)
+			server.waitCloseConn.Done()
+		}()
+	}
+}
+
+func (server *ConiksServer) acceptClient(conn net.Conn, handler func(msg []byte) ([]byte, error)) {
+	defer conn.Close()
+	conn.SetDeadline(time.Now().Add(5 * time.Second))
+
+	// handle request
+	buf := make([]byte, 4096)
+	for {
+		n, err := conn.Read(buf)
+		if err != nil {
+			if err != io.EOF {
+				log.Printf("client read %v: %v", conn.RemoteAddr(), err)
+			}
+			return
+		}
+
+		res, err := handler(buf[:n])
+		if err != nil {
```

Comment:
Ok, that's fair. But I think the listener should additionally check for the type of error and alert the server admin about internal errors.